src/device_state_estimator.cpp

```diff
@@ -0,0 +1,257 @@
+      {
+          .t_estim_ = min_delay->t_process_,
+          .stamp_ = min_delay->device_wall_stamp_,
+          .origin_ = overflow_range.compensate(min_delay->t_origin_),
+      };
+  recent_clocks_.push_back(clock);
+  if (recent_clocks_.size() >= CLOCK_SAMPLES)
+  {
+    recent_clocks_.pop_front();
+  }
+
+  if (recent_clocks_.size() <= 1)
+  {
+    scip2::logger::debug()
+        << "initial origin: " << clock.origin_
+        << ", delay: " << min_delay->delay_
+        << ", delay sigma: " << comm_delay_.sigma_
+        << ", device timestamp: " << min_delay->device_wall_stamp_
+        << std::endl;
+
+    return true;
+  }
+
+  std::vector<ClockState> clocks;
+  for (size_t i = 1; i < recent_clocks_.size(); ++i)
+  {
+    for (size_t j = 0; j < i; ++j)
+    {
+      const double t_diff =
+          (recent_clocks_[i].t_estim_ - recent_clocks_[j].t_estim_).toSec();
+      const double origin_diff =
+          (recent_clocks_[i].origin_ - recent_clocks_[j].origin_).toSec();
+      clocks.emplace_back(
+          recent_clocks_[i].origin_,
+          (t_diff - origin_diff) / t_diff,
+          recent_clocks_[i].stamp_);
+    }
+  }
+  std::sort(clocks.begin(), clocks.end());
+  clock_ = clocks[clocks.size() / 2];
+
+  scip2::logger::debug()
+      << "origin: " << clock_.origin_
+      << ", gain: " << clock_.gain_
+      << ", delay: " << min_delay->delay_
+      << ", delay sigma: " << comm_delay_.sigma_
+      << ", device timestamp: " << min_delay->device_wall_stamp_
+      << std::endl;
+
+  return true;
+}
+
+std::vector<SyncSample>::const_iterator Estimator::findMinDelay(const OriginFracPart& overflow_range) const
+{
+  if (sync_samples_.size() == 0)
+  {
+    return sync_samples_.cend();
+  }
+  auto it_min_delay = sync_samples_.cbegin();
+  for (auto it = sync_samples_.cbegin() + 1; it != sync_samples_.cend(); it++)
+  {
+    if (overflow_range.isOnOverflow(it->t_process_))
+    {
+      continue;
+    }
+    if (it->delay_ < it_min_delay->delay_)
+    {
+      it_min_delay = it;
+    }
+  }
+  return it_min_delay;
+}
+
+OriginFracPart Estimator::originFracOverflow() const
+{
+  if (sync_samples_.size() == 0)
+  {
+    return OriginFracPart();
+  }
+
+  const ros::Duration max_delay = comm_delay_.min_ + delaySigma();
+  int valid_samples = 0;
+
+  auto it_min_origin = sync_samples_.begin();
+  auto it_max_origin = sync_samples_.begin();
+  for (auto it = sync_samples_.begin() + 1; it != sync_samples_.end(); it++)
+  {
+    if (it->delay_ > max_delay)
+    {
+      continue;
+    }
+    valid_samples++;
+    if (it->t_origin_ < it_min_origin->t_origin_)
+    {
+      it_min_origin = it;
+    }
+    if (it->t_origin_ > it_max_origin->t_origin_)
+    {
+      it_max_origin = it;
+    }
+  }
+
+  if (valid_samples < MIN_SYNC_SAMPLES ||
+      it_min_origin->delay_ > max_delay ||
+      it_max_origin->delay_ > max_delay)
+  {
+    return OriginFracPart();
+  }
+
+  double t_min = std::fmod(it_min_origin->t_process_.toSec(), 0.001);
```

Comment:
[optional] To avoid typos, we could create a constant for `0.001` as it seems to be used quite often.

Reply:
Defined `DEVICE_TIMESTAMP_RESOLUTION`